Added format parameter to sync command

Freakin opened the pull request:
Signed-off-by: Brendan Aye <brendan.aye@t-mobile.com>

Freakin commented:
PR to fix #1262

TomSweeneyRedHat commented on cmd/skopeo/sync.go:
```diff
@@ -562,6 +580,7 @@ func (opts *syncOptions) run(args []string, stdout io.Writer) error {
 		DestinationCtx:                        destinationCtx,
 		ImageListSelection:                    imageListSelection,
 		OptimizeDestinationImageAlreadyExists: true,
+		ForceManifestMIMEType:                 manifestType,
```
any man page changes?

mtrmac replied:
Also `completions/bash/skopeo` (which seems to be missing most of `sync` options).

Freakin replied:
@mtrmac would adding completions for `sync` (which don't exist today) be out of scope for this PR?

rhatdan replied:
No you should add them.

mtrmac replied:
Please do add a completion for the new `--format` option.

I don’t think it would be fair to block this PR on adding full completion support for the pre-existing options, though we certainly wouldn’t reject that work :)

vrothberg reviewed (commented):
Thanks!

As @TomSweeneyRedHat mentioned, could you add the new option to the `skopeo-sync` man page?  Also, a test or two in `./integration` would be good.  The formatting code paths are well exercised but I want to make sure that we're not silently regressing in the future.

mtrmac reviewed (commented):
\*shrug\* I don’t care strongly about having this option in `sync`, as such.

But looking at the original report

> However, when I attempted to use multi-arch images with `sync`, my registry (Harbor) only uses the tag on the most recently pushed image and the previously pushed/tagged images have their tags cleared.

Do you mean that all tags in the same repo are removed on a multi-arch push? Or that somehow before there were multiple images with the same tag, and those are now cleared? Or are you just saying that after copying image1 to repo:tag and image2 to repo:tag, image1 no longer has the repo:tag label? Something else? Either way, that report is rather concerning.

Is it possible that the difference is something else, maybe using `--all`? Or, unlikely, were you trying to create a multi-arch image by repeatedly copying the various single-arch images to the same tag?

If that described behavior happens (and it’s not just a tag move on a copy of new image to the same tag), it’s very surprising and rather concerning; AFAICS there’s nothing `skopeo sync` specific to that, and it’s not clear how changing the image format should matter. Having `skopeo copy` untag unrelated images seems like a very serious unwanted side effect that should be understood in detail before adding workarounds like this and papering over the situation.

mtrmac commented on docs/skopeo-sync.1.md:
```diff
@@ -54,6 +54,8 @@ Path of the authentication file for the destination registry. Uses path given by
 
 **--dest** _transport_ Destination transport.
 
+**--format, -f** _manifest-type_ Manifest type (oci, v2s1, or v2s2) to use when saving image to directory using the 'dir:' transport (default is manifest type of source)
```
The `dir:` transport restriction is not true.

Freakin commented on cmd/skopeo/sync.go:
```diff
@@ -536,6 +540,20 @@ func (opts *syncOptions) run(args []string, stdout io.Writer) error {
+		default:
+			return fmt.Errorf("unknown format %q. Choose one of the supported formats: 'oci', 'v2s1', or 'v2s2'", opts.format.value)
+		}
+	}
```
> Do you mean that all tags in the same repo are removed on a multi-arch push? Or that somehow before there were multiple images with the same tag, and those are now cleared? Or are you just saying that after copying image1 to repo:tag and image2 to repo:tag, image1 no longer has the repo:tag label? Something else? Either way, that report is rather concerning.

When copying image1 to `repo:tag` and then copying image2 to `repo:tag`, the `tag` is removed from image1

I think this is likely a limitation of Harbor rather than any issue with Skopeo.  When I attempt to push a multi-arch image using `-all` parameter in `skopeo sync` it fails on Harbor < 2.x.  If multiple architectures for the same `image:tag` are pushed sequentially, the above-mentioned behavior is observed.

With Harbor 2.x and their support for OCI manifests, it appears to correctly infer the MIME type and uses an OCI manifest by default to support multi-arch.  


> Please share this in utils.go instead of copy&pasting from the copy subcommand.

Just to clarify, you're suggesting something like a `parseManifestFormat` function in `utils.go` to be used for `copy` and `sync` commands?

mtrmac replied:
> Just to clarify, you're suggesting something like a `parseManifestFormat` function in `utils.go` to be used for `copy` and `sync` commands?

Yes, please. (Alternatively, it could make sense to share much more of the CLI handling for setting up `copy.Options` between `copy` and `sync`, similarly to `imageOptions` providing `imageOptions.newSystemContext`, but that’s way out of scope of this PR.)

Freakin replied:
Moved parsing to `utils.go` and added a test for it

mtrmac commented:
> When copying image1 to `repo:tag` and then copying image2 to `repo:tag`, the `tag` is removed from image1

Thanks. That looks entirely expected in the container registry model: If a client later requests to download `repo:tag`, it will expect to get image2 and never image1; so the `repo:tag` is no longer pointing to image1, i.e. that it was “removed” from image1 in a sense.

So that is unrelated to the need for this option?

---

> When I attempt to push a multi-arch image using `-all` parameter in `skopeo sync` it fails on Harbor < 2.x.

A schema2 image fails by default but it does work with a manual `--format ? How does that fail? The code is supposed to _automatically_ try various formats. If there is a bug in that code, we should fix it if possible, rather than force users to manually use an option.

> With Harbor 2.x and their support for OCI manifests, it appears to correctly infer the MIME type and uses an OCI manifest by default to support multi-arch.

Or are you saying that all multi-arch pushes fail on Harbor < 2.x, and they automatically work without the `--format` option on ≥ 2.x? So the option is not _necessary_ to make sync work?

If the multi-arch code should work by default but doesn’t fall back correctly, please share a `skopeo --debug sync` output (maybe in the issue, not in this PR).

If this option is not required for the sync to work, it might still make sense to provide it, I just want to make sure we are not ignoring a bug that should be fixed.

Freakin commented:
> > When copying image1 to `repo:tag` and then copying image2 to `repo:tag`, the `tag` is removed from image1
> 
> Thanks. That looks entirely expected in the container registry model: If a client later requests to download `repo:tag`, it will expect to get image2 and never image1; so the `repo:tag` is no longer pointing to image1, i.e. that it was “removed” from image1 in a sense.
> 
> So that is unrelated to the need for this option?
> 
> > When I attempt to push a multi-arch image using `-all` parameter in `skopeo sync` it fails on Harbor < 2.x.
> 
> A schema2 image fails by default but it does work with a manual `--format` ? How does that fail? The code is supposed to _automatically_ try various formats. If there is a bug in that code, we should fix it if possible, rather than force users to manually use an option.
> 
> > With Harbor 2.x and their support for OCI manifests, it appears to correctly infer the MIME type and uses an OCI manifest by default to support multi-arch.
> 
> Or are you saying that all multi-arch pushes fail on Harbor < 2.x, and they automatically work without the `--format` option on ≥ 2.x? So the option is not _necessary_ to make sync work?
> 
> If the multi-arch code should work by default but doesn’t fall back correctly, please share a `skopeo --debug sync` output (maybe in the issue, not in this PR).
> 
> If this option is not required for the sync to work, it might still make sense to provide it, I just want to make sure we are not ignoring a bug that should be fixed.

Yeah looking like I'm just an idiot.  Not sure what problem I was having initially with Harbor 2.x, but doing a normal `sync --all` from docker hub to Harbor 2.x works fine, automatically determines OCI manifest format, and uploads the properly tagged multi-arch image without any issues.

Harbor 1.x does not support multi-arch image manifests and fails as expected with `Manifest.list is not supported` response

I'm happy to finish up this PR or it can be closed if you don't see a reason to add the param to `sync`.  Doesn't look like I need it for my purposes as originally thought.

mtrmac commented:
Thanks very much for the update.

> I'm happy to finish up this PR or it can be closed if you don't see a reason to add the param to `sync`.

Up to you, I don’t feel strongly about this at all. As the above shows, sometimes the discussion about the option can be more helpful than the existence of the option :)

OTOH it’s certainly attractive for `copy` and `sync` to behave very similarly and to provide substantially the same tunables.

rhatdan commented:
I like the idea of the option.

Freakin commented:
Added basic integration test for `sync --format`

Let me know if anything else is needed for this PR

rhatdan commented on cmd/skopeo/utils.go:
```diff
@@ -2,17 +2,21 @@ package main
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
+
```
Remove empty line.

rhatdan commented on cmd/skopeo/utils_test.go:
```diff
@@ -4,10 +4,13 @@ import (
 	"github.com/spf13/cobra"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+
```
Remove blank line.

rhatdan commented on integration/sync_test.go:
```diff
@@ -12,8 +12,11 @@ import (
+	"github.com/containers/image/v5/manifest"
 	"github.com/containers/image/v5/types"
 	"github.com/go-check/check"
+
```
Remove blank line.

rhatdan reviewed (approved):
Nice job, a few nits and missing completions.

mtrmac commented:
> Let me know if anything else is needed for this PR

Completions (discussed elsewhere) and an update to `docs/skopeo-sync.1.md`, I think. Code looks good.

Freakin commented:
> Completions (discussed elsewhere) and an update to `docs/skopeo-sync.1.md`, I think. Code looks good.

Added all completions for sync.  

What update is needed for `docs/skopeo-sync.1.md`?  I added the format parameter previously and corrected the language regarding the `dir:` restrictions based on previous feedback.

Freakin commented:
Looks like `validate` is failing on `gofmt` due to wanting to sort imports.  Sorted imports don't look like are being used in the current HEAD of master.

Should I add a commit with sorted imports?

mtrmac commented:
> What update is needed for `docs/skopeo-sync.1.md`? I added the format parameter previously and corrected the language regarding the `dir:` restrictions based on previous feedback.

I’m sorry, I must have mistakenly viewed only one of the commits or something.

mtrmac commented on cmd/skopeo/sync.go:
```diff
@@ -95,6 +96,7 @@ See skopeo-sync(1) for details.
 	flags := cmd.Flags()
 	flags.BoolVar(&opts.removeSignatures, "remove-signatures", false, "Do not copy signatures from SOURCE images")
 	flags.StringVar(&opts.signByFingerprint, "sign-by", "", "Sign the image using a GPG key with the specified `FINGERPRINT`")
+	flags.VarP(newOptionalStringValue(&opts.format), "format", "f", `MANIFEST TYPE (oci, v2s1, or v2s2) to use when saving image to directory using the 'dir:' transport (default is manifest type of source)`)
```
```suggestion
	flags.VarP(newOptionalStringValue(&opts.format), "format", "f", `MANIFEST TYPE (oci, v2s1, or v2s2) to use when saving image (default is manifest type of source)`)
```

or maybe without the (default is…) part because the code conversions automatically, and explaining that in `--help` would be too verbose.

Same in the man page.

mtrmac commented:
> Looks like `validate` is failing on `gofmt` due to wanting to sort imports. Sorted imports don't look like are being used in the current HEAD of master.

The failure says
> `Please reformat the above files using "gofmt -s -w" and commit the result.`

and that should be all that is required. (The imports are sorted by package name, not purely line-by-line, and only within the empty-line-delineated sections. So removing an empty line can change the order.)

> Should I add a commit with sorted imports?

It might be easier to squash all the commits into one, including the changes made by the `gofmt` step.

Freakin commented:
> > Looks like `validate` is failing on `gofmt` due to wanting to sort imports. Sorted imports don't look like are being used in the current HEAD of master.
> 
> The failure says
> 
> > `Please reformat the above files using "gofmt -s -w" and commit the result.`
> 
> and that should be all that is required. (The imports are sorted by package name, not purely line-by-line, and only within the empty-line-delineated sections. So removing an empty line can change the order.)
> 
> > Should I add a commit with sorted imports?
> 
> It might be easier to squash all the commits into one, including the changes made by the `gofmt` step.

Fixed docs and help language and squashed commits

rhatdan commented:
LGTM

TomSweeneyRedHat commented:
LGTM
@vrothberg ?

vrothberg reviewed (commented):
LGTM, thanks for contributing!
